### ISSUE TITLE
AttributeError: 'module' object has no attribute 'PKCS5_SALT_LEN'

### ISSUE THREAD

**Issue by masterkorp** [saltstack/salt#22851]
I was trying to setup a [development environment](http://docs.saltstack.com/en/latest/topics/development/hacking.html) and got stuck when trying to run the tests.

After setting the environment, I've tried to [run the tests](http://docs.saltstack.com/en/latest/topics/development/tests/index.html) with:
```
pip install -r requirements/dev_python27.txt
./setup.py test
```

But  ``./setup test`` gives me the following error:
```
$ ./setup.py test
2015.2.0-1064-g3b17858
running test
running build
running build_py
running build_scripts
running test
Traceback (most recent call last):
  File "/path/to/my/repo/salt/tests/runtests.py", line 16, in <module>
    from integration import TestDaemon, TMP  # pylint: disable=W0403
  File "/path/to/my/repo/salt/tests/integration/__init__.py", line 54, in <module>
    import salt.config
  File "/path/to/my/repo/salt/salt/config.py", line 39, in <module>
    import salt.utils.sdb
  File "/path/to/my/repo/salt/salt/utils/sdb.py", line 9, in <module>
    import salt.loader
  File "/path/to/my/repo/salt/salt/loader.py", line 23, in <module>
    import salt.utils.event
  File "/path/to/my/repo/salt/salt/utils/event.py", line 79, in <module>
    import salt.payload
  File "/path/to/my/repo/salt/salt/payload.py", line 17, in <module>
    import salt.crypt
  File "/path/to/my/repo/salt/salt/crypt.py", line 22, in <module>
    from M2Crypto import RSA, EVP
  File "/path/to/my/repo/salt/playground/lib/python2.7/site-packages/M2Crypto/__init__.py", line 24, in <module>
    import ASN1
  File "/path/to/my/repo/salt/playground/lib/python2.7/site-packages/M2Crypto/ASN1.py", line 12, in <module>
    import BIO
  File "/path/to/my/repo/salt/playground/lib/python2.7/site-packages/M2Crypto/BIO.py", line 221, in <module>
    class CipherStream(BIO):
  File "/path/to/my/repo/salt/playground/lib/python2.7/site-packages/M2Crypto/BIO.py", line 227, in CipherStream
    SALT_LEN = m2.PKCS5_SALT_LEN
AttributeError: 'module' object has no attribute 'PKCS5_SALT_LEN'
```

I am on Arch Linux, using salt 3b178580525c680d467322caae43f87766c43bf3 from the develop branch.

**Comment by jfindlay** [comment]
@masterkorp, this looks like you're missing an m2crypto-related dependency.  The documentation on running the tests needs to be updated.  Usually I do something like this to prepare a system to run the tests:

```yaml
#/etc/salt/master 
fileserver_backend:
  - git
  - roots
gitfs_remotes:
  - git://github.com/saltstack/salt-jenkins.git
```
```yaml
#/etc/salt/minion 
master: localhost
```
```console
# salt local-minion state.sls git.salt
```

**Comment by masterkorp** [comment]
Hello @jfindlay 

I've followed your instructions, the problem persists:
```
$ salt local-minion state.sls git.salt
Traceback (most recent call last):
  File "/path/to/my/repo/salt/playground/bin/salt", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/path/to/my/repo/salt/scripts/salt", line 10, in <module>
    salt_main()
  File "/path/to/my/repo/salt/salt/scripts.py", line 341, in salt_main
    import salt.cli.salt
  File "/path/to/my/repo/salt/salt/cli/salt.py", line 10, in <module>
    from salt.utils import parsers, print_cli
  File "/path/to/my/repo/salt/salt/utils/parsers.py", line 26, in <module>
    import salt.config as config
  File "/path/to/my/repo/salt/salt/config.py", line 39, in <module>
    import salt.utils.sdb
  File "/path/to/my/repo/salt/salt/utils/sdb.py", line 9, in <module>
    import salt.loader
  File "/path/to/my/repo/salt/salt/loader.py", line 23, in <module>
    import salt.utils.event
  File "/path/to/my/repo/salt/salt/utils/event.py", line 79, in <module>
    import salt.payload
  File "/path/to/my/repo/salt/salt/payload.py", line 17, in <module>
    import salt.crypt
  File "/path/to/my/repo/salt/salt/crypt.py", line 22, in <module>
    from M2Crypto import RSA, EVP
  File "/path/to/my/repo/salt/playground/lib/python2.7/site-packages/M2Crypto/__init__.py", line 24, in <module>
    import ASN1
  File "/path/to/my/repo/salt/playground/lib/python2.7/site-packages/M2Crypto/ASN1.py", line 12, in <module>
    import BIO
  File "/path/to/my/repo/salt/playground/lib/python2.7/site-packages/M2Crypto/BIO.py", line 221, in <module>
    class CipherStream(BIO):
  File "/path/to/my/repo/salt/playground/lib/python2.7/site-packages/M2Crypto/BIO.py", line 227, in CipherStream
    SALT_LEN = m2.PKCS5_SALT_LEN
AttributeError: 'module' object has no attribute 'PKCS5_SALT_LEN'
(playground)
```

Any suggestions would be welcome.

**Comment by jfindlay** [comment]
@rallytime, do you have any ideas?

**Comment by masterkorp** [comment]
After a small resarch it seems that the problem resisdes on https://github.com/swig/swig/issues/344 as the M2Crypto module breaks( refhttps://github.com/M2Crypto/M2Crypto/issues/24)

The workaround right now its to downgrade to swig ``3.0.4``

**Comment by jfindlay** [comment]
@masterkorp, thanks for your work on this.

**Comment by masterkorp** [comment]
ping

**Comment by jleroy** [comment]
I avoided this bug by installing @mtrmac patched version of M2Crypto (not yet merged): 
 ```
pip uninstall M2Crypto
pip install git+https://github.com/mtrmac/M2Crypto.git@swig-3.0.5-22.3#egg=M2Crypto
```

**Comment by jfindlay** [comment]
Also, m2crypto has been deprecated as of Beryllium, which should be out in a month or two.

**Comment by dalepotter** [comment]
Thanks for the update @jfindlay - is the advice to wait for a month or two for this?

At the moment, I'm unable to install Salt master on OSX.

@jleroy Unfortunatley your suggested pip install resulted in:
```
Command "/usr/bin/python -c "import setuptools, tokenize;__file__='/private/var/folders/rr/1dkyvc_510dg69jb1b4qfbvm0000gn/T/pip-build-kV72_m/M2Crypto/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /var/folders/rr/1dkyvc_510dg69jb1b4qfbvm0000gn/T/pip-AKA71S-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /private/var/folders/rr/1dkyvc_510dg69jb1b4qfbvm0000gn/T/pip-build-kV72_m/M2Crypto
```

**Comment by dalepotter** [comment]
Update - in the end I seem to have got Salt working on OSX 10.10

`sudo pip uninstall M2Crypto`

Download 'M2Crypto-0.21.1-py2.7-macosx-10.8-intel.egg' from http://chandlerproject.org/Projects/MeTooCrypto#Downloads

`sudo easy_install M2Crypto-0.21.1-py2.7-macosx-10.8-intel.egg`

From there `sudo salt-master --log-level=all` shows no errors.

**Comment by jfindlay** [comment]
@masterkorp, are you still encountering this issue on Arch?  @dalepotter, I am interested to know if 2015.8.0rc2 works for you, `pip install salt==2015.8.0rc2`.

**Comment by terminalmage** [comment]
@masterkorp Actually, it looks like this is an m2crypto issue after all.

See the discussion [here](https://github.com/martinpaljak/M2Crypto/issues/60) and [here](https://gitlab.com/m2crypto/m2crypto/issues/47).

The issue was improper swig usage on the part of the M2Crypto developers. This should be fixed soon with the release of 0.22.4.

In the meantime, downgrading swig to 3.0.4 will allow M2Crypto 0.22.3 to build.

Since this is an upstream bug, and M2Crypto is no longer used in Salt 2015.8.0 and later, I am going to close this.
